### PR TITLE
K8SPXC-341 : haproxy tests

### PR DESCRIPTION
**Pull request by AlyHKafoury**
[K8SPXC-341](https://jira.percona.com/browse/K8SPXC-341)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| scaling | passed |
| init-deploy | passed |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | failed |
| upgrade | passed |
| limits | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| monitoring | failed |
| upgrade-consistency | failed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| big-data | passed |
| self-healing | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | failed |
| scaling | passed |
| scaling-proxysql | passed |
| upgrade | passed |
| upgrade-consistency | failed |
| security-context | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | failed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| limits | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | failed |
| upgrade | failed |
| demand-backup | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| upgrade-consistency | failed |
| operator-self-healing | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| affinity | passed |
| one-pod | passed |
| demand-backup-encrypted-with-tls | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| big-data | passed |
| recreate | failed |
| scaling | passed |
| restore-to-encrypted-cluster | failed |
| self-healing | failed |
| scaling-proxysql | passed |
| limits | passed |
| demand-backup | passed |
| monitoring | passed |
| upgrade | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| upgrade-consistency | failed |
| monitoring-2-0 | passed |
| security-context | passed |
| scheduled-backup | failed |
| affinity | passed |
| demand-backup-encrypted-with-tls | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| init-deploy | passed |
| scaling | passed |
| recreate | passed |
| storage | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| restore-to-encrypted-cluster | failed |
| upgrade | passed |
| monitoring | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| upgrade-consistency | failed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by cap1984** [comment]
Please add test to Jenkinsfile

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| scaling | passed |
| big-data | passed |
| recreate | passed |
| self-healing | failed |
| limits | passed |
| scaling-proxysql | passed |
| monitoring | passed |
| self-healing-advanced | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | failed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| upgrade-consistency | failed |
| operator-self-healing | passed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| recreate | passed |
| storage | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| limits | passed |
| restore-to-encrypted-cluster | failed |
| upgrade | passed |
| self-healing-advanced | passed |
| monitoring | failed |
| upgrade-consistency | passed |
| demand-backup | failed |
| security-context | passed |
| monitoring-2-0 | failed |
| affinity | passed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | failed |

File: e2e-tests/haproxy/run [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+set -o errexit
+set -o xtrace
+
+test_dir=$(realpath $(dirname $0))
+. ${test_dir}/../functions
+
+check_haproxy_writer() {
+  for i in $(seq 0 2); do
+    local haproxy_pod_ip=$(kubectl_bin describe pod $cluster-haproxy-$i | grep IP | head -n1 | awk '{print $2}')
```
Comment (hors) [inline e2e-tests/haproxy/run:11]:
```suggestion
    local haproxy_pod_ip=$(kubectl_bin get pods some-name-haproxy-0 -o jsonpath='{.status.podIP}') 
```

**Review by hors — CHANGES_REQUESTED** [review]

**Review by cap1984 — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| recreate | passed |
| init-deploy | passed |
| scaling | passed |
| storage | passed |
| self-healing | passed |
| limits | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| security-context | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | passed |
| demand-backup-encrypted-with-tls | failed |

File: e2e-tests/haproxy/conf/some-name.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
+apiVersion: pxc.percona.com/v1
```
Comment (delgod) [inline e2e-tests/haproxy/conf/some-name.yml:1]: please rename it to `haproxy` (similar to other tests)
reason: each unique configuration has its own name

File: e2e-tests/haproxy/run [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  desc 'check advanced options are enabled in haproxy statefulset'
+  compare_kubectl pdb/$cluster-haproxy
+  compare_kubectl statefulset/$cluster-haproxy
+
+  desc 'enable proxy-sql'
+  apply_config "$test_dir/conf/$cluster-proxysql.yml"
+  sleep 50
+  kubectl_bin get statefulset some-name-proxysql
+  kubectl_bin get svc some-name-proxysql
```
Comment (delgod) [inline e2e-tests/haproxy/run:41]:
```suggestion
compare_kubectl statefulset/some-name-proxysql
compare_kubectl service/some-name-proxysql
```

File: e2e-tests/haproxy/run [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  desc 're-enable haproxy'
+  apply_config "$test_dir/conf/$cluster.yml"
+  sleep 50
+  compare_kubectl statefulset/$cluster-haproxy
+  kubectl_bin get svc some-name-haproxy
```
Comment (delgod) [inline e2e-tests/haproxy/run:47]:
```suggestion
compare_kubectl service/some-name-haproxy
```

File: e2e-tests/functions [hunk 5]
```diff
@@ -561,20 +564,29 @@ spinup_pxc() {
     apply_config "$config"
 
     desc 'check if all 3 Pods started'
-    wait_for_running "$cluster-proxysql" 1
+    if [ $haproxy = true ]; then
+        wait_for_running "$cluster-haproxy" 3
+    else
+        wait_for_running "$cluster-proxysql" 1
+    fi
```
Comment (delgod) [inline e2e-tests/functions:571]: please rebase this changes on brach from PR https://github.com/percona/percona-xtradb-cluster-operator/pull/524

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| storage | passed |
| init-deploy | passed |
| big-data | passed |
| scaling | passed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| init-deploy | passed |
| big-data | passed |
| storage | passed |
| scaling | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| storage | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| limits | passed |
| monitoring | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| upgrade-consistency | passed |
| security-context | passed |
| affinity | passed |
| demand-backup | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| haproxy | failed |

File: e2e-tests/functions [hunk 6]
```diff
@@ -585,6 +586,11 @@ spinup_pxc() {
     if [[ "$IMAGE_PXC" =~ 5\.7 ]] && [[ "$(is_keyring_plugin_in_use "$cluster")" ]]; then
         encrypt='ENCRYPTION=\"Y\"'
     fi
+    if [ $haproxy = true ]; then
+        local servicename="$cluster-haproxy"
+    else
+        local servicename="$cluster-proxysql"
+    fi
```
Comment (delgod) [inline e2e-tests/functions:593]: haproxy and servicename vars are not used, so we don't need this change anymore

File: e2e-tests/haproxy/run [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  apply_config "$test_dir/conf/$cluster-proxysql.yml"
+  sleep 50
+  compare_kubectl statefulset/some-name-proxysql
+  compare_kubectl service/some-name-proxysql
+
+  desc 're-enable haproxy'
+  apply_config "$test_dir/conf/$cluster.yml"
+  sleep 50
+  compare_kubectl statefulset/$cluster-haproxy
+  compare_kubectl service/some-name-haproxy
```
Comment (delgod) [inline e2e-tests/haproxy/run:47]: minor improvement of style
```suggestion
  compare_kubectl statefulset/$cluster-proxysql
  compare_kubectl service/$cluster-proxysql

  desc 're-enable haproxy'
  apply_config "$test_dir/conf/$cluster.yml"
  sleep 50
  compare_kubectl statefulset/$cluster-haproxy
  compare_kubectl service/$cluster-haproxy
```

File: e2e-tests/haproxy/conf/some-name.yml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    image: -haproxy
+    affinity:
+      antiAffinityTopologyKey: "kubernetes.io/hostname"
+    tolerations:
+    - key: "node.alpha.kubernetes.io/unreachable"
+      operator: "Exists"
+      effect: "NoExecute"
+      tolerationSeconds: 6000
+    podDisruptionBudget:
+      maxUnavailable: 1
```
Comment (delgod) [inline e2e-tests/haproxy/conf/some-name.yml:42]: minor improvement to test if it passed
```suggestion
      maxUnavailable: 2
```

File: e2e-tests/haproxy/run [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (delgod) [inline e2e-tests/haproxy/run:48]: can we also check podDisruptionBudget
```suggestion
    compare_kubectl pdb/$cluster-haproxy
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| storage | passed |
| recreate | failed |
| haproxy | failed |
| big-data | passed |
| self-healing | failed |
| init-deploy | failed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| scaling-proxysql | failed |
| demand-backup | passed |
| monitoring | failed |
| operator-self-healing | passed |
| upgrade | passed |
| monitoring-2-0 | failed |
| upgrade-consistency | passed |
| scheduled-backup | failed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| demand-backup-encrypted-with-tls | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| haproxy | passed |
| recreate | failed |
| scaling | failed |
| big-data | passed |
| init-deploy | failed |
| self-healing | failed |
| scaling-proxysql | failed |
| limits | passed |
| restore-to-encrypted-cluster | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| upgrade | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| haproxy | passed |
| big-data | passed |
| recreate | failed |
| scaling | failed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | failed |
| limits | passed |
| restore-to-encrypted-cluster | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| affinity | passed |
| security-context | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| recreate | failed |
| storage | passed |
| haproxy | failed |
| big-data | passed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | failed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| monitoring | failed |
| demand-backup | passed |
| self-healing-advanced | passed |
| upgrade | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| security-context | passed |
| affinity | passed |
| scheduled-backup | failed |
| one-pod | passed |
| auto-tuning | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| haproxy | failed |
| storage | passed |
| recreate | failed |
| scaling | failed |
| self-healing | failed |
| init-deploy | failed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | failed |
| limits | passed |
| self-healing-advanced | passed |
| upgrade | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| monitoring | failed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Pull request merged**